### PR TITLE
7903455: JMH: Add "mempool" profiler

### PR DESCRIPTION
**Pull request by ericcaspole**
I started off to show only the code cache related size, but I don't see any reason to not print all the available mbean pool info. We often adapt app or library code to run as JMH, to use all its benefits. In those cases seeing the code cache stats is more interesting than say, for a regular JDK API JMH benchmark. I verified the CodeHeap stats match well with +PrintCodeCache. There is mbean pool data for all the GCs as well.

Example output:

Benchmark                                                              (iterations)  Mode  Cnt      Score   Error  Units
SpillCode.testSpillForManyInts                                                   10  avgt    5    705.433 ± 9.570  ns/op
SpillCode.testSpillForManyInts:·CodeHeap 'non-nmethods'.used                     10  avgt    5   1139.000             kB
SpillCode.testSpillForManyInts:·CodeHeap 'non-profiled nmethods'.used            10  avgt    5    209.750             kB
SpillCode.testSpillForManyInts:·CodeHeap 'profiled nmethods'.used                10  avgt    5    924.875             kB
SpillCode.testSpillForManyInts:·Compressed Class Space.used                      10  avgt    5    403.375             kB
SpillCode.testSpillForManyInts:·G1 Eden Space.used                               10  avgt    5   8192.000             kB
SpillCode.testSpillForManyInts:·G1 Old Gen.used                                  10  avgt    5   8160.000             kB
SpillCode.testSpillForManyInts:·G1 Survivor Space.used                           10  avgt    5        ≈ 0             kB
SpillCode.testSpillForManyInts:·Metaspace.used                                   10  avgt    5   3617.320             kB
SpillCode.testSpillForManyInts:·Total CodeHeap.used                              10  avgt    5   2267.000             kB
SpillCode.testSpillForManyInts:·Total MemPools.used                              10  avgt    5  22639.695             kB

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [CODETOOLS-7903455](https://bugs.openjdk.org/browse/CODETOOLS-7903455): JMH: Add "mempool" profiler

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jmh.git pull/99/head:pull/99` \
`$ git checkout pull/99`

Update a local copy of the PR: \
`$ git checkout pull/99` \
`$ git pull https://git.openjdk.org/jmh.git pull/99/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 99`

View PR using the GUI difftool: \
`$ git pr show -t 99`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jmh/pull/99.diff">https://git.openjdk.org/jmh/pull/99.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jmh/pull/99#issuecomment-1508788038)

### CODE REVIEW

**Review by shipilev — CHANGES_REQUESTED** [review]
This looks useful! 

I think the whole thing can be simpler, though:

```
public class MemPoolProfiler implements InternalProfiler {
    static final double BYTES_PER_KIB = 1024D;

    @Override
    public String getDescription() {
        return "Memory pool/footprint profiling via standard MBeans";
    }

    @Override
    public void beforeIteration(BenchmarkParams benchmarkParams, IterationParams iterationParams) {
        for (MemoryPoolMXBean b : ManagementFactory.getMemoryPoolMXBeans()) {
            b.resetPeakUsage();
        }
    }

    @Override
    public Collection<? extends Result> afterIteration(BenchmarkParams benchmarkParams, IterationParams iterationParams, IterationResult result) {
        List<Result> results = new ArrayList<>();

        long sumCodeHeap = 0L;
        long sum = 0L;
        for (MemoryPoolMXBean bean : ManagementFactory.getMemoryPoolMXBeans()) {
            long used = bean.getPeakUsage().getUsed();
            if (bean.getName().contains("CodeHeap")) {
                sumCodeHeap += used;
            }
            sum += used;

            results.add(new ScalarResult(Defaults.PREFIX + "mempool." + bean.getName() + ".used", used / BYTES_PER_KIB, "KiB", AggregationPolicy.MAX));
        }

        results.add(new ScalarResult(Defaults.PREFIX + "mempool.total.codeheap.used", sumCodeHeap / BYTES_PER_KIB, "KiB", AggregationPolicy.MAX));
        results.add(new ScalarResult(Defaults.PREFIX + "mempool.total.used", sum / BYTES_PER_KIB, "KiB", AggregationPolicy.MAX));
        return results;
    }
}
```

In addition to this version being more readable, it also does only a single pass through the beans, which means sums and individual metrics would always agree.

**Comment by shipilev** [comment]
Please also make sure to pass GHA tests: go to https://github.com/ericcaspole/jmh/actions and enable the workflows there.

**Comment by ericcaspole** [comment]
Thanks for your changes, it is better with your version.

File: jmh-core/src/main/java/org/openjdk/jmh/profile/MemPoolProfiler.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            sum += used;
+
+            results.add(new ScalarResult(Defaults.PREFIX + "mempool." + bean.getName() + ".used", used / BYTES_PER_KIB, "KiB", AggregationPolicy.MAX));
+        }
+
+        results.add(new ScalarResult(Defaults.PREFIX + "mempool.total.codeheap.used", sumCodeHeap / BYTES_PER_KIB, "KiB", AggregationPolicy.MAX));
+        results.add(new ScalarResult(Defaults.PREFIX + "mempool.total.used", sum / BYTES_PER_KIB, "KiB", AggregationPolicy.MAX));
+        return results;
+    }
+}
```
Comment (shipilev) [inline jmh-core/src/main/java/org/openjdk/jmh/profile/MemPoolProfiler.java:72]: Need a new line at the end of file.

**Review by shipilev — APPROVED** [review]
Looks good, feel free to push once GHA are clean.
(If we need something else from this profiler, we can do that separately)

**Pull request closed without merging**